### PR TITLE
Adding filestore support

### PR DESCRIPTION
**Pull request by chiayi**
Add Filestore support to Jupyterhub solutions template.
The decision was to have Filestore be a shared storage solution within Jupyterhub, meaning that all users will share a Filestore instance.
Added flag `enable_filestore_use` for enabling Filestore use.

### CODE REVIEW

**Comment by chiayi** [comment]
@saikat-royc adding you here since for some reason I can't seem to add you as reviewer.

File: jupyter-on-gke/README.md [hunk 1]
```diff
@@ -130,6 +130,8 @@ Currently there are 2 choices for storage:
 
 2. GCSFuse - `gcsfuse.csi.storage.gke.io` uses GCS Buckets and require users to pre-create buckets with name format `gcsfuse-{username}`
 
+3. Filestore - `filestore.csi.storage.gke.io` uses Filestore autoprovisioning to create persistent volumes. Uses `Retain` reclaim policy
```
Comment (saikat-royc) [inline jupyter-on-gke/README.md:133]: ... uses CSI Dynamic provisioning to provision Filestore based k8s volumes

File: jupyter-on-gke/jupyter_config/config-filestore.yaml [hunk 2]
```diff
@@ -0,0 +1,167 @@
+# Copyright 2023 Google LLC
```
Comment (saikat-royc) [inline jupyter-on-gke/jupyter_config/config-filestore.yaml:1]: 2024

File: jupyter-on-gke/README.md (hunk 1 above)
Comment (saikat-royc) [reply jupyter-on-gke/README.md:133]: also please update  L127 to 3 storage choices

File: jupyter-on-gke/persistent_storage_deployments/filestore-pvc.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: filestore-claim
+  namespace: <NAMESPACE>
+spec:
+  accessModes:
+    - ReadWriteMany
+  storageClassName: ""
```
Comment (saikat-royc) [inline jupyter-on-gke/persistent_storage_deployments/filestore-pvc.yaml:9]: storageClassName: filestore-storage-class

(which is created in the filestore-storage-class.yaml)

File: jupyter-on-gke/storage.md [hunk 4]
```diff
@@ -9,3 +9,40 @@ GCSFuse allow users to mount GCS Buckets as their local filesystem. This option
 ![Profiles Page](images/gcs_bucket.png)
 
 Since this bucket in GCS, there is built in permission control and access outside of the clutser.
+
+## Filestore
+
+Filestore has built in auto provisioning. This means that the first time this option is picked for a users, it will auto provision a Filestore instance then mount it to the Jupyter instance.
+
+**Important Note:** Currently the teir used is the `standard` tier. Different other teirs may require different handling.
```
Comment (saikat-royc) [inline jupyter-on-gke/storage.md:17]: typo: tiers

Comment (saikat-royc) [reply jupyter-on-gke/storage.md:17]: Also it is important to point to the filestore storageclass network param. If using a non default n/w for the GKE cluster, the user needs to point to the right network (https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/filestore-csi-driver#storage-class) https://screenshot.googleplex.com/gdZujwkZ2Wu6MR5

File: jupyter-on-gke/storage.md [hunk 5]
```diff
@@ -9,3 +9,40 @@ GCSFuse allow users to mount GCS Buckets as their local filesystem. This option
...
+
+Filestore can be accessed by any GCE instance within the same VPC network.
```
Comment (saikat-royc) [inline jupyter-on-gke/storage.md:19]: Also it is important to point to the filestore storageclass network param. If using a non default n/w for the GKE cluster, the user needs to point to the right network (https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/filestore-csi-driver#storage-class) https://screenshot.googleplex.com/gdZujwkZ2Wu6MR5

File: jupyter-on-gke/storage.md [hunk 6]
```diff
@@ -9,3 +9,40 @@ GCSFuse allow users to mount GCS Buckets as their local filesystem. This option
...
+
+### Remounting a Filestore instance
+
+A new Filestore instance will be created everytime the option is selected unless there is already an existing PVC under the format `cliam-{USERNAME}`. This means that to reuse an existing Filestore instance will require manual remounting.
+
+1. Open up the Filestore page in [cloud console](https://console.cloud.google.com/filestore/instances). It will look something like this:
+
+![Filestore instance](images/filestore_instance_screenshot.png)
+
+2. Fill out both `filestire-pv.yaml` and `filestore-pvc.yaml` under the `persistent_storage_deployments` directory. The cloud console will have the necessary information.
```
Comment (saikat-royc) [inline jupyter-on-gke/storage.md:44]: typo: filestore

File: jupyter-on-gke/storage.md [hunk 7]
```diff
@@ -9,3 +9,40 @@ GCSFuse allow users to mount GCS Buckets as their local filesystem. This option
 ![Profiles Page](images/gcs_bucket.png)
 
 Since this bucket in GCS, there is built in permission control and access outside of the clutser.
+
+## Filestore
+
+Filestore has built in auto provisioning. This means that the first time this option is picked for a users, it will auto provision a Filestore instance then mount it to the Jupyter instance.
```
Comment (saikat-royc) [inline jupyter-on-gke/storage.md:15]: [Filestore CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/filestore-csi-driver#storage-class) supports automatic creation of Filestore instances using the CSI dynamic provisioning workflow.

File: jupyter-on-gke/storage.md [hunk 8]
```diff
@@ -9,3 +9,40 @@ GCSFuse allow users to mount GCS Buckets as their local filesystem. This option
...
+        dynamic:
+            pvcNameTemplate: claim-{username}
+...
+```
+
+This means that every single users will have their own PVC+PV. For users that want to share the PV within the same namespace, they can set the `pvcNameTemplate` to a more generic/static template. **Important** Changing the `pvcNameTemplate` means that when re-mounting, the name of the PVC will also have the match this new template.
+
+### Remounting a Filestore instance
+
+A new Filestore instance will be created everytime the option is selected unless there is already an existing PVC under the format `cliam-{USERNAME}`. This means that to reuse an existing Filestore instance will require manual remounting.
```
Comment (saikat-royc) [inline jupyter-on-gke/storage.md:38]: typo: cliam

**Comment by saikat-royc** [comment]
overall looks good with minor comments

**Comment by chiayi** [comment]
I just realized that we are no longer using kubectl with terraform and are instead using helm to install. Going to work on changing it to helm.

File: applications/jupyter/storage.md [hunk 9]
```diff
@@ -9,3 +9,42 @@ GCSFuse allow users to mount GCS Buckets as their local filesystem. This option
...
+
+1. Open up the Filestore page in [cloud console](https://console.cloud.google.com/filestore/instances). It will look something like this:
+
+![Filestore instance](images/filestore_instance_screenshot.png)
+
+2. Fill out both `filestore-pv.yaml` and `filestore-pvc.yaml` under the `persistent_storage_deployments` directory. The cloud console will have the necessary information.
+
+3. Deploy the PVC/PV into the correct namespace (PV is cluster resource so this only applies to PVC)
+
+4. Login to the user and select Filestore as persistent storage.
```
Comment (saikat-royc) [inline applications/jupyter/storage.md:50]: Can you clarify what you mean by "Login to the user"

File: modules/jupyter/jupyter_config/config-filestore.yaml [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    iam.gke.io/gke-metadata-server-enabled: "true"
+# Used for GCSFuse to set the ephemeral storage as the home directory. If not set, it will show a permission error on the pod log when using GCSFuse. 
+  extraEnv:
+    JUPYTER_ALLOW_INSECURE_WRITES: "true"
+  image:
+    name: jupyter/tensorflow-notebook
+    tag: python-3.10
+  startTimeout: 1000
+  extraAnnotations:
+    gke-gcsfuse/volumes: "true"
```
Comment (saikat-royc) [inline modules/jupyter/jupyter_config/config-filestore.yaml:89]: these are not needed if we are not mounting a bucket to the pod

File: modules/jupyter_storage/charts/templates/filestore-pv.yaml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  storageClassName: ""
+  capacity:
+    storage: 1Ti
+  accessModes:
+    - ReadWriteMany
+  persistentVolumeReclaimPolicy: Retain
+  volumeMode: Filesystem
+  csi:
+    driver: filestore.csi.storage.gke.io
+    volumeHandle: "modeInstance/{{ .Values.filestore.remount.location }}/{{ .Values.filestore.remount.instanceName }}/{{ .Values.filestore.remount.sharedName }}"
```
Comment (saikat-royc) [inline modules/jupyter_storage/charts/templates/filestore-pv.yaml:29]: shareName

File: modules/jupyter_storage/charts/values.yaml [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+namespace: ${namespace}
+
+filestore:
+  tier: standard
+  network: default
+  remount:
+    enabled: false
+    location: us-central1
+    instanceName: "xyz-555-abc"
+    sharedName: "shared-name-xyz"
```
Comment (saikat-royc) [inline modules/jupyter_storage/charts/values.yaml:28]: share name

File: applications/jupyter/storage.md (hunk 9 above)
Comment (chiayi) [reply applications/jupyter/storage.md]: I elaborated on it:  "Login to Jupyterhub using the desired user...."

Comment (saikat-royc) [reply applications/jupyter/storage.md]: Ack

**Comment by spencer-p** [comment]
Closing old issues with merge conflicts. Please rebase and re-open if still relevant.

**Pull request closed without merging**